aten/src/ATen/native/Batching.cpp

```diff
@@ -20,6 +21,73 @@ static bool has_level(const Tensor& self, int64_t level) {
   return it != bdims.end();
 }
 
+// Returns a Tensor with batch dim with level `level` turned into a regular dimension,
+// as well as a logical dim index of where said dimension is in the returned tensor.
+//
+// The reason why we want to return the index of where said dimension is in the returned
+// tensor is because we want to keep track of which dimension used to be the batch
+// dimension so that we can move it to the correct logical dimension specified by
+// `out_dims` in vmap. For example, if we had
+// >>> x = torch.randn(2, 3, 5)
+// >>> vmap(lambda x: x, in_dims=0, out_dims=1)(x)
+// then right when we are about to exit the vmap block, x is a BatchedTensor with a
+// batch dimension at (physical) index 0. Note that the batch dimension doesn't
+// always have to exist at (physical) index 0. When we undo the batch dimension,
+// we want to move it to dimension 1 (as specified by out_dims). So we return the
+// index at which the batch dim appears so that we can move it to the correct place.
+// later down the line.
+static std::pair<Tensor,int64_t> remove_existing_batch_dim(const Tensor& self, int64_t level) {
+  const auto* batched = maybeGetBatched(self);
+  TORCH_INTERNAL_ASSERT(batched != nullptr);
+  auto bdims = batched->bdims();
+  if (bdims.size() == 1) {
+    TORCH_INTERNAL_ASSERT(bdims[0].level() == level);
+    return std::make_pair(batched->value(), bdims[0].dim());
```

Comment:
Crumbs, I've gotten confused again. Is the dim() of a BatchDim logical or physical? I read the comment but it didn't help me understand:

```
// a BatchDim represents a "private" dimension on a Tensor created inside of
// vmap. It is a (level, dim) tuple, with the `dim` indicating which dimension
// is being vmap'ed over and the `level` being an identifier for which vmap
// said dimension was created inside.
```

Reply:
I guess, based on 

```
// The batch dimensions are treated as being "private"; they are not user-visible.
// For example, in the following Tensor,
//    bt = BatchedTensorImpl(ones(2, 3, 5, 7), [(lvl=1, dim=0), (lvl=2, dim=1)])
// dimensions 0 and 1 are batch dimensions.

```

this means the dim is physical.

Reply:
OK, so in this case it's ok to pun the physical dim into a logical dim because you've eliminated all of the batch dimensions in this case. :+1:

Reply:
dim() of a BatchDim is physical. I'll go beef up the documentation for BatchDim

Reply:
Added a note to `BatchDim` that the dimension index is physical